### ISSUE TITLE
Culling management

### ISSUE THREAD

**Issue by Bertram25** [OpenDungeons/OpenDungeons#170]
Atm, the creature and tile culling management is done in the same named class, through a culling manager that updates optionally within the camera manager.

Unfortunately, the code is making the game crash and is very hard to grasp. I don't blame @tomluchowski at all for that, I rather understand why he couldn't finish it since it's very hard to follow.

To get out of this without getting crazy, I'd propose we reuse some simple culling algorithm given by Ogre and test around this.
You can find a simple culling handler here, for instance that could be tested:
http://www.ogre3d.org/tikiwiki/tiki-index.php?page=Frustum+Culling+In+Object+Space

The base code source is public domain, as stated in the page's bottom.

Nothing urgent but we will come back at this sooner or later.

**Comment by oyvindln** [comment]
As no one has taken up the task, I suggest we drop the custom culling code for now. It hasn't been updated since we changed to the server-client structure, so it's probably broken anyhow, and it's contributing to build time and coupling between classes.

**Comment by Danimal696** [comment]
i would like to ask you to implement a culling manager nevertheless, no matter how basic in the future, its a very usefull piece of code

**Comment by hwoarangmy** [comment]
> I suggest we drop the custom culling code for now

+1. As you say, there are probably many things changed since the server-client structure that can be simplified/improved. Moreover, there are many calls (especially on rendering) that avoids making changes even if they are not used.
And finally, IMHO, when we will have a better idea of how the FOW will be implemented, it will probably help us on finding a better way to activate culling.

**Comment by Bertram25** [comment]
I do think we can drop it, too. As for the FOW, I also think that it will help a lot in term of performance once in, as well. Since it will somehow "hide" everything not in view of the creatures per default, after all.

I'm also sure the culling proposed in the description will be easier to add once the former is removed.

**Comment by akien-mga** [comment]
That might sound like a strange request after all this time where culling has been evoked, but I still have no clue about what it means :o) Could someone explain me the concept briefly? ;-)

**Comment by hwoarangmy** [comment]
> Could someone explain me the concept briefly?

Well, the concept is not very complex. The idea is that when you play a game, you look through the camera at only a part of the whole area. For example, while you look at your dungeon, you don't see the 3 other players' dungeons.
The idea of culling is to not render what you are not looking at. If we take our previous example, if you look at your dungeon, the other players dungeons won't be rendered.

But even if the concept is pretty understandable, it is not easy to do...

**Comment by akien-mga** [comment]
Thanks :-) Sounds important indeed, but I also think we can drop the current code if all three of you coders think it's not a good start.

**Comment by Bertram25** [comment]
> Thanks :-) Sounds important indeed, but I also think we can drop the current code if all three of you coders think it's not a good start.

Well, the current code looks quite complicated while I don't claim to have the solution. ;)

**Comment by oyvindln** [comment]
Ogre already does culling. It may be possible to fine tune how the culling works to achieve better performance, but I don't think it's worth bothering with at this stage.

**Comment by hwoarangmy** [comment]
> I don't think it's worth bothering with at this stage

Yep, that's what I would call "early optimization". Lets first wait for a problem to try to solve it ^^

**Comment by Bertram25** [comment]
Ok, then I guess it's settled. The issue can be closed once the experimental culling management has been removed.

> i would like to ask you to implement a culling manager nevertheless, no matter how basic in the future, its a very usefull piece of code

@Danimal696 While it's always useful to cull stuff you don't need on screen, it will bring only troubles if we make things more complicated than needed. IMHO, this is the kind of stuff that can wait for after the Fog of War. Ogre 2.0 will also help in that area AFAIK.

**Comment by tomluchowski** [comment]
What's wrong with the culling for creatures ? 
There are like two mechanism for TIles and for Creatures , the former is by default disabled . 
No generic  algorithm would be ever as fast as hand written one , as teh clothes  from TESCO would fit better than those from  cobbler ... 
I 'm having hard times in installing CEGUI and seeing what's currentlu in the development branch ....

**Comment by akien-mga** [comment]
@tomluchowski Regarding your CEGUI issue, as discussing on IRC, you should:
1) Build cegui from source and make sure to enable the freetype2 dependency. You can check my package for a working example of dependencies that produce a _good_ CEGUI: https://svnweb.mageia.org/packages/cauldron/cegui/current/SPECS/cegui.spec?view=markup
2) Report a bug at openSUSE so that they add pkgconfig(freetype2) as a build dependency for their CEGUI package to enable freetype2 support (you can also point them to above spec file btw, contrarily to *SUSE we don't put frightening copyright notices on 30 lines of script at Mageia ;-) ).

**Comment by Bertram25** [comment]
@tomluchowski Hey there. Nice to see you're still around.
I do think you should first concentrate on remaking your build environment. That may take time but you will have a clearer view of where we are atm.

Speaking of the culling management, yes there are two algorithms, and yes custom ones are fasters, but only when working, right?
I can understand @oyvindln 's proposal to remove some code that is not following code guidelines, looks no-op AFAIK, and seems very hard to follow.
But since you're around, I beg you to explain us in detail how the two algo work, what files they are using, where you stopped concerning the implementation, the problems you were trying to solve etc...

Why? Because it seems nobody is atm willing to dive in that part as it looks like a time-consuming one without a clear hope to see it fixed or ready to be enabled, notably in parallel of the will to implement a fog of war, I guess sooner than later.

That said, if we can get some of your expertise on this, we could earn a lot of time on understanding what is there and why, and maybe make better decisions about the removal or the fix of this part.

Best regards,

**Comment by tomluchowski** [comment]
_Edited for readability_

> But since you're around, I beg you to explain us in detail how the two algo work, what files they are using, where you stopped concerning the implementation, the problems you were trying to solve etc...

Could you start from there and point what you need to know more with  :? 
http://forum.freegamedev.net/viewtopic.php?t=5480

**Comment by Bertram25** [comment]
An interesting read about Ogre 2.0 regarding the current state of the culling management:
http://www.ogre3d.org/forums/viewtopic.php?f=25&t=82178

IMHO, what is reassuring is that the culling management in Ogre 2.0 seems already way better than in 1.x with still some place for improvement of course.
